src/metabase/lib/js.cljs

```diff
@@ -234,6 +234,31 @@
   [a-query]
   (lib.core/drop-empty-stages a-query))
 
+(defn ^:export as-returned
+  "When a query has aggregations in stage `N`, there's an important difference between adding an expression to stage `N`
+  (with access to the colums before aggregation) or adding it to stage `N+1` (with access to the aggregations and
+  breakouts).
+
+  Given `a-query` and `stage-number`, this returns a **JS object** with `query` and `stageIndex` keys, for working with
+  \"what it returns\". If there is already a later stage, that stage is reused. Appends a new stage if we were already
+  looking at the last stage.
+
+  > **Code health:** Healthy"
+  [a-query stage-number]
+  (if (empty? (lib.core/aggregations a-query stage-number))
+    ;; No extra stage needed with no aggregations.
+    #js {:query      a-query
+         :stageIndex stage-number}
+    ;; An extra stage is needed, so see if one already exists.
+    (if-let [next-stage (->> (lib.util/canonical-stage-index a-query stage-number)
+                             (lib.util/next-stage-number a-query))]
+      ;; Already an extra stage, so use it.
+      #js {:query      a-query
+           :stageIndex next-stage}
+      ;; No new stage, so append one.
+      #js {:query      (lib.core/append-stage a-query)
+           :stageIndex -1})))
```

Comment:
I wonder if returning a relative index (-1) when the input was an absolute one can trip callers. If they are in a context where they work with absolute indexes and calculate stage indexes themselves, then they might `inc` this index and get surprised when it points the first stage.